### PR TITLE
[OBS-2059] Dump shard map on SIGUSR1

### PR DESCRIPTION
**Pull request by golfdish**
TODO:
- [ ] don't hardcode output path

Example input statsrelay.json:
```json
{"statsd":
 {
   "bind": "0.0.0.0:8127",
   "validate": false,
   "tcp_cork": false,
   "shard_map": [ "10.0.1.75:8128",
                  "10.0.1.75:8128",
                  "10.0.1.75:8128",
                  "10.0.1.75:8128"
                  ,
                  "10.0.12.27:8128",
                  "10.0.12.27:8128",
                  "10.0.12.27:8128",
                  "10.0.12.27:8128"
                  ,
                  "10.0.13.49:8128",
                  "10.0.13.49:8128",
                  "10.0.13.49:8128",
                  "10.0.13.49:8128"
                  ,
                  "10.0.13.51:8128",
                  "10.0.13.51:8128",
                  "10.0.13.51:8128",
                  "10.0.13.51:8128"
                  ,
                  "10.0.132.197:8128",
                  "10.0.132.197:8128",
                  "10.0.132.197:8128",
                  "10.0.132.197:8128"
                  ,
                  "10.0.134.167:8128",
                  "10.0.134.167:8128",
                  "10.0.134.167:8128",
                  "10.0.134.167:8128"
                  ,
                  "10.0.139.249:8128",
                  "10.0.139.249:8128",
                  "10.0.139.249:8128",
                  "10.0.139.249:8128"
                  ,
                  "10.0.144.32:8128",
                  "10.0.144.32:8128",
                  "10.0.144.32:8128",
                  "10.0.144.32:8128"
                  ,
                  "10.0.147.200:8128",
                  "10.0.147.200:8128",
                  "10.0.147.200:8128",
                  "10.0.147.200:8128"
                  ,
                  "10.0.156.157:8128",
                  "10.0.156.157:8128",
                  "10.0.156.157:8128",
                  "10.0.156.157:8128"
                  ,
                  "10.0.18.92:8128",
                  "10.0.18.92:8128",
                  "10.0.18.92:8128",
                  "10.0.18.92:8128"
                  ,
                  "10.0.21.188:8128",
                  "10.0.21.188:8128",
                  "10.0.21.188:8128",
                  "10.0.21.188:8128"
                  ,
                  "10.0.24.65:8128",
                  "10.0.24.65:8128",
                  "10.0.24.65:8128",
                  "10.0.24.65:8128"
                  ,
                  "10.0.26.116:8128",
                  "10.0.26.116:8128",
                  "10.0.26.116:8128",
                  "10.0.26.116:8128"
                  ,
                  "10.0.26.18:8128",
                  "10.0.26.18:8128",
                  "10.0.26.18:8128",
                  "10.0.26.18:8128"
                  ,
                  "10.0.28.122:8128",
                  "10.0.28.122:8128",
                  "10.0.28.122:8128",
                  "10.0.28.122:8128"
                  ,
                  "10.0.28.179:8128",
                  "10.0.28.179:8128",
                  "10.0.28.179:8128",
                  "10.0.28.179:8128"
                  ,
                  "10.0.28.180:8128",
                  "10.0.28.180:8128",
                  "10.0.28.180:8128",
                  "10.0.28.180:8128"
                  ,
                  "10.0.32.38:8128",
                  "10.0.32.38:8128",
                  "10.0.32.38:8128",
                  "10.0.32.38:8128"
                  ,
                  "10.0.35.165:8128",
                  "10.0.35.165:8128",
                  "10.0.35.165:8128",
                  "10.0.35.165:8128"
                  ,
                  "10.0.7.104:8128",
                  "10.0.7.104:8128",
                  "10.0.7.104:8128",
                  "10.0.7.104:8128"
                  ,
                  "10.0.75.182:8128",
                  "10.0.75.182:8128",
                  "10.0.75.182:8128",
                  "10.0.75.182:8128"
                  ,
                  "10.0.87.75:8128",
                  "10.0.87.75:8128",
                  "10.0.87.75:8128",
                  "10.0.87.75:8128"
                  ,
                  "10.0.93.138:8128",
                  "10.0.93.138:8128",
                  "10.0.93.138:8128",
                  "10.0.93.138:8128"

                ],
    "health_metrics_to": {
          "prefix": "statsd-staging-iad.",
          "shard_map":
  ["statsdmonitor-staging-s1.lyft.net:8125", "statsdmonitor-staging-s2.lyft.net:8125"]
  ,
          "suffix": ".instance.statsd-staging-iad-00cafd7e6e77e6463"
    }
  }
}
```

Example output:
```json
{"shard_map": ["10.0.1.75:8128", "10.0.1.75:8128", "10.0.1.75:8128", "10.0.1.75:8128", "10.0.12.27:8128", "10.0.12.27:8128", "10.0.12.27:8128", "10.0.12.27:8128", "10.0.13.49:8128", "10.0.13.49:8128", "10.0.13.49:8128", "10.0.13.49:8128", "10.0.13.51:8128", "10.0.13.51:8128", "10.0.13.51:8128", "10.0.13.51:8128", "10.0.132.197:8128", "10.0.132.197:8128", "10.0.132.197:8128", "10.0.132.197:8128", "10.0.134.167:8128", "10.0.134.167:8128", "10.0.134.167:8128", "10.0.134.167:8128", "10.0.139.249:8128", "10.0.139.249:8128", "10.0.139.249:8128", "10.0.139.249:8128", "10.0.144.32:8128", "10.0.144.32:8128", "10.0.144.32:8128", "10.0.144.32:8128", "10.0.147.200:8128", "10.0.147.200:8128", "10.0.147.200:8128", "10.0.147.200:8128", "10.0.156.157:8128", "10.0.156.157:8128", "10.0.156.157:8128", "10.0.156.157:8128", "10.0.18.92:8128", "10.0.18.92:8128", "10.0.18.92:8128", "10.0.18.92:8128", "10.0.21.188:8128", "10.0.21.188:8128", "10.0.21.188:8128", "10.0.21.188:8128", "10.0.24.65:8128", "10.0.24.65:8128", "10.0.24.65:8128", "10.0.24.65:8128", "10.0.26.116:8128", "10.0.26.116:8128", "10.0.26.116:8128", "10.0.26.116:8128", "10.0.26.18:8128", "10.0.26.18:8128", "10.0.26.18:8128", "10.0.26.18:8128", "10.0.28.122:8128", "10.0.28.122:8128", "10.0.28.122:8128", "10.0.28.122:8128", "10.0.28.179:8128", "10.0.28.179:8128", "10.0.28.179:8128", "10.0.28.179:8128", "10.0.28.180:8128", "10.0.28.180:8128", "10.0.28.180:8128", "10.0.28.180:8128", "10.0.32.38:8128", "10.0.32.38:8128", "10.0.32.38:8128", "10.0.32.38:8128", "10.0.35.165:8128", "10.0.35.165:8128", "10.0.35.165:8128", "10.0.35.165:8128", "10.0.7.104:8128", "10.0.7.104:8128", "10.0.7.104:8128", "10.0.7.104:8128", "10.0.75.182:8128", "10.0.75.182:8128", "10.0.75.182:8128", "10.0.75.182:8128", "10.0.87.75:8128", "10.0.87.75:8128", "10.0.87.75:8128", "10.0.87.75:8128", "10.0.93.138:8128", "10.0.93.138:8128", "10.0.93.138:8128", "10.0.93.138:8128"]}
```

@lyft/observability

### CODE REVIEW

**Comment by golfdish** [comment]
👀 @Longinus00 @mjchoi

File: src/main.c [hunk 1]
```diff
@@ -60,6 +61,45 @@ static void quick_shutdown(struct ev_loop *loop, ev_signal *w, int revents) {
...
+        stats_error_log("dump_shard_map: fdopen failed");
+        close(fd);
+        return;
+    }
+
+    struct config *cfg = (struct config *) w->data;
+    list_t ring = cfg->statsd_config.ring;
+
+    json_t *out_object = json_object();
+    json_t *shard_map = json_array();
```
Comment (Longinus00) [inline src/main.c]: This is never freed, memory leak?

File: src/main.c [hunk 2]
```diff
@@ -60,6 +61,45 @@ static void quick_shutdown(struct ev_loop *loop, ev_signal *w, int revents) {
...
+
+    FILE *f;
+    int fd;
+    // TODO: no hardcode path
+    if ((fd = open("/tmp/shard_map.json", O_RDWR|O_CREAT, 0644)) == -1) {
+        stats_error_log("dump_shard_map: failed to open or create /tmp/shard_map.json");
+        return;
+    }
+
+    if ((f = fdopen(fd, "r+")) == NULL) {
```
Comment (Longinus00) [inline src/main.c]: Can use fopen(PATH, "w") instead of open + fdopen

File: src/main.c (hunk 1 above)
Comment (Longinus00) [reply src/main.c]: or does the json_delete on out_object recursively delete everything inside of it?

**Comment by yashlyft** [comment]
please make the pr against statsrelay-private.git

**Pull request closed without merging**